### PR TITLE
Initial pass for "easier embedded rust"

### PR DESCRIPTION
**Pull request by jamesmunns**
This needs a lot of links and citations still, but is a mostly complete first thought.

This introduces an "Ideas" folder, for things that I've thought about, but haven't actually done any concrete projects for yet.

@amanda-onevariable feel free to review if you'd like (tho #7 probably has higher priority). I'd appreciate any notes on "you should define this", or if you want to fill in any links, that would be awesome too.

### CODE REVIEW

**Comment by jamesmunns** [comment]
I asked folks in my chat room:

> btw, soliciting opinions:
> 
> * [Anatol]: if you have any written up "experience notes" about stuff that CircuitPython does (easily, well) that Rust doesn't; or
> * [Scott M.]: if you have any written up "hey running rust on top of an RTOS is awesome/sucks because" thoughts
> 
> I'd be interested in them :D

[Anatol]: https://github.com/spookyvision/
[Scott M.]: https://github.com/MabezDev

Answers I got back:

([Anatol] - regarding CircuitPython):

> * REPL (fantastic for playing around with e.g. LED values, or PWM, etc.)
> * auto-reload (exposes flash as mass storage, updating /[code.py](http://code.py/) restarts the app)
>     * this includes a fancy special case for display drivers (any kind really, hub75 as well as LCDs): as soon as you "quit the app" (via C-c), the runtime takes over the display and shows the repl (this is of somewhat limited use unless you have some kind of keyboard connected, but ... just very cool)
> * out of the box starts a ton of USB devices (uart, midi, keyboard, etc. - even has host support but only for teensy)
> * mainly though: cross-architecture HAL + "BSC" (e.g. you can do spi = board.SPI() - that's literally all you need) + "OS" (layout, font rendering, animation etc) with very, very few LOC required)
>
> I think [these examples/cheatsheet things](https://github.com/adafruit/awesome-circuitpython/blob/main/cheatsheet/CircuitPython_Cheatsheet.md) are a good showcase, or [these](https://docs.circuitpython.org/projects/displayio-layout/en/latest/examples.html) (if your specific board has a display you don't even have to initialize it, you get board.DISPLAY for free)
>
> oh, speaking of filesystems! it comes with a vfs! so you can mount an sd card on top of your flash fs
>
> ```python
> import os
>  
> import board
> import sdcardio
> import storage
>  
> sd = sdcardio.SDCard(board.SPI(), board.SD_CS)
> vfs = storage.VfsFat(sd)
> storage.mount(vfs, '/sd')
> os.listdir('/sd')
> ```
>
> I don't think building is such a big problem with Rust, the main advantages I see with circuitpython are just lots and lots of conveniences/helpers that make for very short programs
>
> and fast iteration
>
> I think I'll add one more bullet point, a bit more philosophical: micro/circuitpython carry the spirit of Python's "batteries included" ... so you don't have to go and find crates, add them to Cargo.toml - most stuff is just bundled in, and designed to interoperate (kinda like embedded-hal on steroids)

([Scott M.] - Regarding RTOS usage):

> Some initial thoughts of the top of my head.
> 
> I think running an RTOS (that's not written in Rust) is not worth is unless you are porting the standard library too. If you do port the standard library, it's awesome! If you don't, you just end up using unsafe everywhere. The key points imo are:
> 
> * Anyone who knows Rust can write code for a target that supports STD. Many beginners just want to get stuff done, and don't care about choosing alloc impls, or worrying about efficient static memory usage.
> * Access to more of the amazing Rust ecosystem (we have clients running fully fledged tokio on esp's!)
>
> The not so good, but imo everything in this list is solvable:
> 
> * Usually you have to build the RTOS along with your Rust build, for many developers this is no big deal but noobs just want it to work without installing 50 million deps
>     * Easy solution to this is to precompile std with your RTOS, just like how std is distributed now
> * Sometimes STD is too high level, i.e a filesystem on an embedded system could be flash based, or a full on fat32/exfat impl
>     * In these instances, for esp-idf anyway, there is some setup to do outside of std to allow the fs parts of std to work correctly.
>     * This gap can be filled nicely with supporting crates, we have esp-idf-hal and esp-idf-svc (svc = services) to do that.

File: Ideas/easier-embedded-rust.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    * Tools like Cargo for build + package management
+    * Ability to run `no_std` programs easily
+    * Modern language features (modules, iterators, type system)
+    * Making it "harder to get things wrong" - use the compiler to detect errors statically
+* Building good tools:
+    * Way better debugger tools via probe-rs, and tools built on top like probe-run and probe-rs-cli
+    * Tooling for logging, like defmt
+* Reasonably good docs, and teaching materials
+
+However, Rust does not have as good of a "getting started (new to embedded and/or new to rust" story, or a story for "just want things to work (artists/hobbyists)".
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md]: The first set of parentheses is missing the second ). 
I feel like this sentence could be rephrased to make it flow better generally (?)
- Despite these benefits, Rust does not have as good of a "getting started" story, neither for those new to embedded and/or new to Rust nor for artists/hobbyists who "just want things to work."
- but also "not have as good of" in comparison to some other programming language (plz specify as example)? or just in general, like "Rust doesn't have a good 'getting started' story"

File: Ideas/easier-embedded-rust.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+For whatever shortcomings (performance, reliability, licensing, "under the hood complexity") exist (or are perceived to exist) for ecosystems like Arduino and CircuitPython/MicroPython: they make getting started much easier than Rust does today.
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md]: I think I would rephrase this also for flow -
- Ecosystems like Arduino and CircuitPything/MicroPython make getting started much easier than Rust does today, despite whatever shortcomings which exist (or are perceived to exist) with those ecosystems, such as performance, reliability, licensing or "under the hood complexity."

File: Ideas/easier-embedded-rust.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+## Concepts to consider
+
+Balancing "simple" and "easy". How can we make things *easier*, without hiding all of the complexity under the hood, or making things more complex just so we can hide them under the hood?
+
+Hidden complexity makes it very hard if you ever fall off the "happy path". Requiring "extra" tools not typically used by "production" embedded rust users makes it harder to find help. Special systems means effort is split.
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md]: I think a couple more words here would help with clarity
- harder to find help (in the community? with the tools?)
- Special systems (with these extra, hard to acquire tools?) means effort is split (how?)

File: Ideas/easier-embedded-rust.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+How can this be done efficiently? Arduino and CircuitPython both have corporate backing in addition to their open source efforts. While there is a growing open source embedded rust community, any new approaches would have to leverage existing work or start from scratch.
+
+How can this be done in a way that has a "path" to more traditional Rust usage? When users become more comfortable with embedded Rust, and would like to make lower level modifications, how can they do this without abandoning everything they've learned, or without experiencing a steep learning curve?
+
+## Potential Research Areas
+
+### Better enumeration of Why and How other approaches are easier
+
+It would be good to do an overview of today's Arduino and CircuitPython environments, and see what they provide (or don't) to make the experience easier. It would also be useful to capture where their approaches "fall over", or what limitations typical users run into.
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md:43]: what do you mean by "fall over"?

File: Ideas/easier-embedded-rust.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+This approach may limit application to targets with enough resources (typically RAM and Flash, including external flash) to handle both an "OS"/"Host Environment", as well as the user application.
+
+### Provide a "common enough" base library
+
+This approach is somewhat similar to how Arduino approaches the problem: making code "seem" the same to the end user, by forcing libraries to use a common, high level set of interfaces.
+
+User code is compiled together with selected libraries to produce a single, static code image which is flashed to the device (either via debugger or bootloader).
+
+Arduino does some extra work here by providing it's own IDE and package manager ("board" and "library" managers) to manage dependencies, that may be already available out of the box in Rust.
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md]: its* own

File: Ideas/easier-embedded-rust.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+Although Rust does have language level support for standard interfaces via Traits (rather than 'common header/api' approach used in C++/Arduino) - leveraged heavily by embedded-hal, the "OSS zeitgeist" of Rust is to provide MUCH more control over implementation details that are important to "power users", but less so to "hobbyist/artist" users (at least for initial usage). This is often confusing or even overwhelming for newcomers. This is very commonly seen on "early main code" - setting up clocks, peripherals, and other resources.
```
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md]: This first sentence is quite long, let me see if I can understand/flow it out:
- Although Rust does have language level support for standard interfaces via Traits (rather than the 'common header/api' approach used in C++/Arduino) - leveraged heavily by embedded-hal - the "OSS Zeitgeist" of Rust is to provide MUCH more control over implementation details that are important to "power users", but less so to "hobbyist/artist" users (at least for initial usage).

ok I think the only thing throwing me off a bit is the "- leveraged heavily by embedded-hal -" part, if it is supposed to be a parenthetical statement or.. well, I think this sentence could be hashed out a bit for clarity, or separated into a couple sentences.

File: Ideas/easier-embedded-rust.md
Comment (amanda-onevariable) [inline Ideas/easier-embedded-rust.md:1]: General question: should Rust always be capitalised because it is a name of the language? Embedded is not always a name, but in what context should it be capitalised- when in conjunction with Rust? This also goes for Trait, HAL, Python, etc (uniformity above all)

**Review by amanda-onevariable — COMMENTED** [review]
Oh, is this.. Ok, yeah, I reviewed it, comments in line and general question also somewhere.🥸

File: Ideas/easier-embedded-rust.md (hunk 4 above)
Comment (jamesmunns) [reply Ideas/easier-embedded-rust.md:43]: "fall over" is sort of a euphemism for failing, or when something stops working.

Sometimes you have a system that works, until it doesn't.

For example, if you had to fill out a form every time you shipped something, that might be fine for shipping one or two things a day. But if 100 orders come in one day, and you have to fill out a form by hand for all of those, it'd get overwhelming! This might be a point where the "filling out all forms by hand" approach "falls over", and you would want to switch to something like having a computer fill out the form for you.

File: Ideas/easier-embedded-rust.md
Comment (jamesmunns) [reply Ideas/easier-embedded-rust.md:1]: Generally yes! I don't have a style guide, but it's probably good to be consistent.

**Comment by jamesmunns** [comment]
@amanda-onevariable I'm going to go ahead and merge this, addressed most of the comments you made, if you think of any improvements or "todo" items I should capture, please feel free to add them in a PR or open an issue to track them :)
